base/special/trig.jl

```diff
@@ -1105,19 +1160,93 @@ function _cosc(x::Number)
 #  Float64.([(-1)^n*big(pi)^(2n)/((2n+1)*factorial(2n-1)) for n = 1:6])
-_cosc(x::Union{Float64,ComplexF64}) =
+_coscpi(x::Union{Float64,ComplexF64}) =
     fastabs(x) < 0.14 ? x*evalpoly(x^2, (-3.289868133696453, 3.2469697011334144, -1.1445109447325053, 0.2091827825412384, -0.023460810354558236, 0.001781145516372852)) :
-    isinf_real(x) ? zero(x) : ((pi*x)*cospi(x)-sinpi(x))/((pi*x)*x)
-_cosc(x::Union{Float32,ComplexF32}) =
+    isinf_real(x) ? zero(x) : ((s,c)=sincospi(x); ((pi*x)*c-s)/((pi*x)*x))
+_coscpi(x::Union{Float32,ComplexF32}) =
     fastabs(x) < 0.26f0 ? x*evalpoly(x^2, (-3.289868f0, 3.2469697f0, -1.144511f0, 0.20918278f0)) :
-    isinf_real(x) ? zero(x) : ((pi*x)*cospi(x)-sinpi(x))/((pi*x)*x)
-_cosc(x::Float16) = Float16(_cosc(Float32(x)))
-_cosc(x::ComplexF16) = ComplexF16(_cosc(ComplexF32(x)))
+    isinf_real(x) ? zero(x) : ((s,c)=sincospi(x); ((pi*x)*c-s)/((pi*x)*x))
+_coscpi(x::Float16) = Float16(_coscpi(Float32(x)))
+_coscpi(x::ComplexF16) = ComplexF16(_coscpi(ComplexF32(x)))
+
+"""
+    sincu(x)
+
+Compute the un-normalized ``\\mathrm{sinc}`` function ``\\sin(x) / x`` if ``x \\neq
+0``, and ``1`` if ``x = 0``.
+
+The normalization is with respect to integration over all real values of ``x``:
+```math
+\\int_{-\\infty}^{\\infty} \\mathrm{sincu}(x) \\, \\mathrm{d}x = \\pi.
+```
+
+See also [`coscu`](@ref), the derivative of this function, or [`sinc`](@ref),
+[`cosc`](@ref), [`sincpi`](@ref), or [`coscpi`](@ref) for different normalizations.
+
+!!! compat "Julia 1.11"
+    This method was added in Julia 1.11.
+"""
+sincu(x::Number) = _sincu(float(x))
+_sincu(x::Number) = iszero(x) ? one(x) : isinf_real(x) ? zero(x) : sin(x)/x
+@inline _sincu(x::Union{Float64,ComplexF64}) =
+    fastabs(x) < 0.0031 ? evalpoly(x^2, (1.0, -0.16666666666666666, 0.008333333333333333)) :
+    isinf_real(x) ? zero(x) : sin(x)/x
+@inline _sincu(x::Union{Float32,ComplexF32}) =
+    fastabs(x) < 0.1571f0 ? evalpoly(x^2, (1.0f0, -0.16666667f0, 0.008333334f0)) :
+    isinf_real(x) ? zero(x) : sin(x)/x
+_sincu(x::Float16) = Float16(_sincu(Float32(x)))
+_sincu(x::ComplexF16) = ComplexF16(_sincu(ComplexF32(x)))
+
+"""
+    coscu(x)
+
+Compute ``\\cos(x) / x - \\sin(x) / x^2`` if ``x \\neq 0``, and ``0`` if ``x = 0``. This is
+the derivative of [`sincu`](@ref).
+
+See also [`sinc`](@ref), [`cosc`](@ref), [`sincpi`](@ref), or [`coscpi`](@ref) for different
+normalizations.
+
+!!! compat "Julia 1.11"
+    This method was added in Julia 1.11.
+"""
+coscu(x::Number) = _coscu(float(x))
+function _coscu(x::Number)
+    # naive coscu formula is susceptible to catastrophic
+    # cancellation error near x=0, so we use the Taylor series
+    # for small enough |x|.
+    if fastabs(x) < 1.57
+        # generic Taylor series: ∑ (-1)^n (x)^{2n-1}/a(n) where
+        # a(n) = (1+2n)*(2n-1)! (= OEIS A174549)
+        s = (term = -x)/3
+        x² = term^2
+        ε = eps(fastabs(term)) # error threshold to stop sum
+        n = 1
+        while true
```

Comment:
Is this really the best implimentation?

Reply:
Well, given that there are explicit versions for `Float16` (kinda), `Float32`, and `Float64`, this is basically a fallback.  But I think it is a pretty good one.  (Looks like it's due to @stevengj.)  I did actually do some experimentation with `BigFloat` and `Double64` to see how well it works, and it does pretty darn well.  The only notable thing I found is that we could potentially approximate the number of terms to add [`ceil((0.865*(x1 - 1) / log(x1) - (x1 * 0.0339)) + 1)` is a surprisingly good estimate for sincpi], but I don't think we'd really gain much, and it may not be robust.  The only questionable thing is the choice of 0.5π as the cutoff, and basically I find that it doesn't make much difference.